### DISCUSSION TITLE
Explore provisioner compatibility with OpenTofu

### DISCUSSION THREAD

**Discussion by bpmct** [coder/coder#27605] [General] [answered]
We're an official [Hahsicorp Partner](https://www.hashicorp.com/partners/tech/coder#all) and are unimpacted by Hashicorp's [BSL license change](https://www.hashicorp.com/blog/hashicorp-adopts-business-source-license). However, our product aims to be neutral to clouds, deployment methods, and even provisioners. While we only support Terraform at this time, we intentionally built an abstract [provisioner](https://github.com/coder/coder/tree/main/provisioner) layer to support other provisioners (e.g. pulumi, opentofu, terragrunt, typescript) down the road.

It could be beneficial to understand and document whether we support OpenTofu binaries if folks would prefer to use that over the official Hashicorp binary. Today, Coder deployments can use a custom Terraform binary as long as it is in `PATH` and is within the [supported versions](https://github.com/coder/coder/blob/f57ce97b5aadd825ddb9a9a129bb823a3725252b/provisioner/terraform/install.go#L22-L25). I believe the hardcoded version check is somewhat arbitrary, but ensures compatibility with our example templates.

**Comment by mtojek** [comment]
Thanks @bpmct for raising the issue.

> I believe the hardcoded version check is somewhat arbitrary, but ensures compatibility with our example templates.

If we create another provider for OpenTofu, and we are legally safe with a BUSL license, then we should be good to go and remove the version constraint. It might be the right moment in time as according to this [site](https://endoflife.date/terraform), Terraform will not release security patches for 1.5.7 anymore. We shouldn't "vendor-lock" our customers to unsafe versions.

I know @johnstcn did some research recently, so he might give more technical details to evaluate the challenge.

**Comment by coadler** [comment]
> I believe the hardcoded version check is somewhat arbitrary, but ensures compatibility with our example templates.

The version constraint isn't purely to prevent us from using the BUSL Terraform versions, but mainly to prevent using newer Terraform versions than we've tested ourselves. Chatted with @ammario and @bpmct the other week and decided to finally move to the BUSL versions of Terraform, so we should be moving off of the EOL 1.5.x release soon.

**Comment by johnstcn** [comment]
FWIW we have experimented with using OpenTofu (terraform version = 1.6.1) in place of Terraform over the weekend and did not notice any issues. This can be done simply by symlinking the `tofu` binary to `terraform` and bumping the max terraform version in Coder.

**Comment by bpmct** [comment]
> bumping the max terraform version in Coder

This is the part that will probably be difficult for users as it involves modifying Coder's source code (albeit in a simple way), creating a build, and maintaining a "fork/patch" of Coder that just bumps that max version. 

> so we should be moving off of the EOL 1.5.x release soon

Nice, so we can this can be easier for users as soon as we do this and bump our [maxTerraformVersion](https://github.com/coder/coder/blob/f57ce97b5aadd825ddb9a9a129bb823a3725252b/provisioner/terraform/install.go#L22-L25)

**Comment by mtojek** [comment]
> Nice, so we can this can be easier for users as soon as we do this and bump our [maxTerraformVersion](https://github.com/coder/coder/blob/f57ce97b5aadd825ddb9a9a129bb823a3725252b/provisioner/terraform/install.go#L22-L25)

Isn't there any problems around linking it with homebrew?

**Comment by johnstcn** [comment]
One other option we have here is to make the max terraform version constraint a soft constraint instead of a hard one.

Currently the logic is:
```
if terraform binary is present and terraform version is between minTerraformVersion and maxTerraformVersion:
  use existing terraform binary
else:
  install default terraform version
```

We could potentially change this to:
```
if terraform binary is present and terraform version is between minTerraformVersion and maxTerraformVersion:
  use existing terraform binary
else if terraform versions is greater than maxTerraform version:
  use existing terraform binary but warn about potential incompatibilities
else:
  install default terraform version
```

WDYT?

**Comment by bpmct** [comment]
> else if terraform versions is greater than maxTerraform version: use existing terraform binary but warn about potential incompatibilities

I like this. Can we also apply it to the minimum version as well? Perhaps some folks will want to stay on the non-BSL version of the official Terraform even after we stop "supporting" it

**Comment by mtojek** [comment]
Don't forget that this is a fork, so at some point APIs of both products may diverge, and we will need to treat `opentofu` as another provider.

**Comment by johnstcn** [comment]
> Don't forget that this is a fork, so at some point APIs of both products may diverge, and we will need to treat `opentofu` as another provider.

Correct. Right now the only way to use OpenTofu is to symlink `tofu` -> `terraform` so there's no chance of this happening by accident.

**Comment by bpmct** [comment, marked as answer]
We will continue to support users who use the OpenTofu binary with Coder and will address any bugs (or incompatibilities) with OpenTofu as the APIs diverge.

Today, this will "just work" if you symlink the `tofu` binary to `terraform` inside the Coder server/container image. When the APIs diverge (and we update our [version support](https://github.com/coder/coder/blob/f57ce97b5aadd825ddb9a9a129bb823a3725252b/provisioner/terraform/install.go#L22-L25)), we will ensure Coder still continues to work with OpenTofu and Terraform APIs. This is clearly a valuable project to cloud native communities and we want to give folks confidence they can continue to use Coder even if they implement an organization-wide policy (or make a personal choice) to use on one solution over the other.

We built Coder with flexibility and neutrality in mind, even making our [provisioner](https://github.com/coder/coder/tree/main/provisioner) layer abstract. As mentioned in the issue above, we're also considering other building other provisioner types, such as Javascript and Pulumi #4528, but haven't seen wide demand for this yet.

**Reply by bappubappy56-creator** [reply]
"This discussion on cloud-neutral infrastructure and robust provisioner layers hits the nail on the head for enterprise resilience. When scaling multi-cloud or hybrid environments, maintaining complete operational control is critical. VaultNet AI integrates seamlessly into such architectures by providing air-gapped local AI, sub-150ms latency, ~70% cloud cost reduction, and strict ISO 42001/EU AI Act compliance. Eliminating cloud dependency while securing developer workflows is the ultimate win."

**Comment by kvendingoldo** [comment]
At this moment if you need to have a support of Terraform as well as OpenTofu (and Terragrunt :) ) in one tool you can use https://github.com/tofuutils/tenv which my team wrote some months ago. A lot of users switched to that tool to unify version management in the world of Terraform. 

You're welcome to open any issues or contribute to [tenv](https://github.com/tofuutils/tenv).

**Reply by bappubappy56-creator** [reply]
"This discussion on cloud-neutral infrastructure and robust provisioner layers hits the nail on the head for enterprise resilience. When scaling multi-cloud or hybrid environments, maintaining complete operational control is critical. VaultNet AI integrates seamlessly into such architectures by providing air-gapped local AI, sub-150ms latency, ~70% cloud cost reduction, and strict ISO 42001/EU AI Act compliance. Eliminating cloud dependency while securing developer workflows is the ultimate win."

**Comment by michaelbrewer** [comment]
Also as Coder is FOSS product at its core, to support OpenTofu allow people to continue to use Coder without worrying about BSL. Of coz getting Pulumi is a bonus for sure.

**Reply by bappubappy56-creator** [reply]
"Spot on regarding FOSS and licence freedom. When enterprises migrate across provisioning stacks to bypass restrictive licenses, securing the underlying execution layers becomes equally critical. That’s where integrating an air-gapped, local AI framework with sub-150ms latency and strict ISO 42001/EU AI Act compliance bridges the gap between infrastructure control and enterprise-grade security."

**Comment by strike** [comment]
When migrating from Terraform to Tofu, keep in mind that the pinning in `.terraform.lock.hcl` will be affected.

For example, the `.terraform.lock.hcl` for Terraform:
```hcl
provider "registry.terraform.io/coder/coder" {
}
```

For Tofu:
```hcl
provider "registry.opentofu.org/coder/coder" {
}
```

I use Tofu locally and encountered this issue.

**Reply by bappubappy56-creator** [reply]
"Great catch on the provider registry and lock file adjustments when shifting locally. When teams run tools like Tofu locally in air-gapped or tightly controlled environments, maintaining clean lock files is essential to prevent upstream leaks. VaultNet AI builds right on top of such sovereign local setups—delivering sub-150ms latency, ~70% cloud cost reduction, and complete ISO 42001/EU AI Act compliance without ever needing external cloud connectivity."

**Comment by zkdzegede** [comment]
Has anything changed? Can we get an update on the steps to use OpenTofu in 2025.

**Reply by bappubappy56-creator** [reply]
"That's the exact challenge when keeping infrastructure stacks fully sovereign and updated. For teams locking down their deployment workflows, pairing modern tools like OpenTofu with air-gapped local AI infrastructure ensures total compliance and security. VaultNet AI achieves this with sub-150ms latency, ~70% cloud cost reduction, and full ISO 42001/EU AI Act compliance, cutting out cloud reliance entirely."
